### PR TITLE
[nibeheatpump] Added device support for SMO40 control

### PR DESCRIPTION
**Pull request by MarkusGafner**
Added the support of a new heatpump. Therefore all files got created and modified, as described in the repository, to communicate with this device. Tested with the system in the last 2 months without any problems

### CODE REVIEW

**Comment by fwolter** [comment]
@paulianttila can you take a look?

@MarkusGafner can you fix the sign-off in your commit? You could click on the DCO check details below to see what's wrong and how to fix it.

File: bundles/org.openhab.binding.nibeheatpump/README.md [hunk 1]
```diff
@@ -30,6 +30,9 @@ This binding supports direct serial port connection (RS-485 adapter needed) to h
 | f470-serial     | Serial port connected F470 Heat Pumps            |
 | f470-udp        | UDP connected Nibe F470 Heat Pumps               |
 | f470-simulator  | Simulator for Nibe F470 Heat Pumps               |
+| smo40-serial    | Serial port connected SMO40 controler            |
```
Comment (paulianttila) [inline bundles/org.openhab.binding.nibeheatpump/README.md:33]: Should it be controller?

**Review by paulianttila — CHANGES_REQUESTED** [review]
Just one typo, otherwise LGTM

File: bundles/org.openhab.binding.nibeheatpump/README.md (hunk 1 above)
Comment (MarkusGafner) [reply bundles/org.openhab.binding.nibeheatpump/README.md:33]: Yes of course

File: bundles/org.openhab.binding.nibeheatpump/README.md [hunk 2]
```diff
@@ -1896,3 +1899,12 @@ This binding currently supports following channels for F1x55 pump models:
 | 49380           | Switch    | 0           | 1          | Setting | External ERS 3 accessory bypass at heat                  |                                                                                                                                                                    |
 | 49381           | Switch    | 0           | 1          | Setting | External ERS 2 accessory bypass at heat                  |                                                                                                                                                                    |
 | 49430           | Number    | 0           | 255        | Setting | AUX ERS Fire Place Guard                                 |                                                                                                                                                                    |
+
+
+### SMO40
+
+To keep this documantation light, all parameters are documented in the NIBE ModbusManager except of:
```
Comment (fwolter) [inline bundles/org.openhab.binding.nibeheatpump/README.md:1906]:
```suggestion
To keep this documentation light, all parameters are documented in the NIBE ModbusManager except of:
```

File: bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40-types.xml [hunk 3]
```diff
@@ -0,0 +1,9437 @@
...
+		<label>CL12-BT51 Pool 2 Temp</label>
+		<description>Pool temperature for Pool 2, BT51</description>
+		<state pattern="%.1f °C" readOnly="true">
+		</state>
+	</channel-type>
+	<channel-type id="smo40-40121" advanced="true">
+		<item-type>Number</item-type>
+		<label>BT63 Add Supply Temp</label>
+		<description></description>
+		<state pattern="%.1f §C" readOnly="true">
```
Comment (fwolter) [inline bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40-types.xml:207]:
```suggestion
		<state pattern="%.1f °C" readOnly="true">
```

File: bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40-types.xml [hunk 4]
```diff
@@ -0,0 +1,9437 @@
...
+		<description>Indicates the status of the relays on the ERS accessory. The information is binary encoded. b0: K1, b1:
+			K2, b2: K3, b3: K4</description>
+		<state pattern="%d" readOnly="true">
+		</state>
+	</channel-type>
+	<channel-type id="smo40-40311" advanced="true">
+		<item-type>Number</item-type>
+		<label>External ERS 1 accessory GQ2 speed</label>
+		<description>Indicates the speed of the GQ2 fan speed on the ERS accessory.</description>
+		<state pattern="%d %" readOnly="true">
```
Comment (fwolter) [inline bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40-types.xml:509]: This needs to be escaped. Please check all.
```suggestion
		<state pattern="%d %%" readOnly="true">
```

File: bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40-types.xml [hunk 5]
```diff
@@ -0,0 +1,9437 @@
...
+	<channel-type id="smo40-48375" advanced="true">
+		<item-type>Switch</item-type>
+		<label>Op mode circ.pump Heat Slave 8</label>
+		<description></description>
+		<state readOnly="false"></state>
+	</channel-type>
+	<channel-type id="smo40-48376" advanced="true">
+		<item-type>Switch</item-type>
+		<label>Op mode circ.pump Heat Slave 7</label>
+		<description></descr
```
Comment (fwolter) [inline bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40-types.xml:8110]: Are the spaces in the pattern intentional?

Comment (fwolter) [inline bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40-types.xml:8725]: Is this file generated code? Actually labels are expected to be as short as possible. Guideline is 2-3 words with up to 25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

File: bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml [hunk 6]
```diff
@@ -0,0 +1,168 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="nibeheatpump"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="smo40-udp">
+		<label>UDP Connected Nibe SMO40 Heat Pumps</label>
```
Comment (fwolter) [inline bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml:8]: Labels are expected to be as short as possible. Guideline is 2-3 words with up to 25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

File: bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+		<description></description>
+
+		<channel-groups>
+			<channel-group typeId="smo40-sensor-group-channels" id="sensor"/>
+			<channel-group typeId="smo40-setting-group-channels" id="setting"/>
+		</channel-groups>
+
+		<config-description>
+			<parameter name="hostName" type="text" required="true">
+				<label>Host Name</label>
```
Comment (fwolter) [inline bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml:18]: You could specify `<context>network-address</context>` to get a free validation.

File: bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+				<label>UDP Port for Read Commands</label>
+				<description>UDP port to send read commands to the NibeGW.</description>
+				<default>9999</default>
+			</parameter>
+			<parameter name="writeCommandsPort" type="integer">
+				<label>UDP Port for Write Commands</label>
+				<description>UDP port to send write commands to the NibeGW.</description>
+				<default>10000</default>
+			</parameter>
+			<parameter name="refreshInterval" type="integer">
```
Comment (fwolter) [inline bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml:36]:
```suggestion
			<parameter name="refreshInterval" type="integer" min="0" unit="s">
```

File: bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+		<description></description>
+
+		<channel-groups>
+			<channel-group typeId="smo40-sensor-group-channels" id="sensor"/>
+			<channel-group typeId="smo40-setting-group-channels" id="setting"/>
+		</channel-groups>
+
+		<config-description>
+			<parameter name="serialPort" type="text" required="true">
+				<label>Serial Port</label>
```
Comment (fwolter) [inline bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml:79]: You could specify `<context>serial-port</context>` to get a free validation.

File: bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+				<description>Serial port to connect to the heat pump.</description>
+			</parameter>
+			<parameter name="refreshInterval" type="integer">
```
Comment (fwolter) [inline bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml:82]:
```suggestion
			<parameter name="refreshInterval" type="integer" min="0" unit="s">
```

File: bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+				<default>false</default>
+			</parameter>
+			<parameter name="enableWriteCommands" type="boolean">
+				<label>Enable Write Commands</label>
+				<description>Enable write commands to change heat pump settings. This is experimental feature, use it at your own
+					risk!</description>
+				<default>false</default>
+			</parameter>
+			<parameter name="enableWriteCommandsToRegisters" type="text">
+				<label>Register List for Write Commands</label>
```
Comment (fwolter) [inline bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml:100]: See above. Label length.

File: bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+		<label>Simulator for Nibe SMO40 Heat Pumps</label>
+		<description></description>
+
+		<channel-groups>
+			<channel-group typeId="smo40-sensor-group-channels" id="sensor"/>
+			<channel-group typeId="smo40-setting-group-channels" id="setting"/>
+		</channel-groups>
+
+		<config-description>
+			<parameter name="refreshInterval" type="integer">
```
Comment (fwolter) [inline bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml:138]:
```suggestion
			<parameter name="refreshInterval" type="integer" min="0" unit="s">
```

**Review by fwolter — CHANGES_REQUESTED** [review]
@paulianttila Thanks for reviewing in advance!

@MarkusGafner Thanks for your contribution! Here comes my review feedback.

File: bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40-types.xml (hunk 5 above)
Comment (MarkusGafner) [reply bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40-types.xml:8110]: The label string is given by the manufacturer

Comment (MarkusGafner) [reply bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40-types.xml:8725]: As written before, the strings are given by the manufacturer and I won't change them.

Comment (fwolter) [reply bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40-types.xml:8110]: Seems like GitHub mixed up some lines in this review. I was refering to line 8110:
`<state min="-10" max="10" step="1" pattern="%d  " readOnly="false">`

**Comment by fwolter** [comment]
Your changes look good! Can you fix the DCO check? You can click on details below to view the necessary commands.

File: bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml (hunk 7 above)
Comment (MarkusGafner) [reply bundles/org.openhab.binding.nibeheatpump/src/main/resources/OH-INF/thing/SMO40.xml:18]: The code of SMO40.xml is equal to the code uf d1x45.xml and 1x55.xml.
I suggest to change all classes and keep them equal in their behaviour

**Review by fwolter — APPROVED** [review]
LGTM

As OH3 is already feature-freezed, your PR will be merged as soon as OH3 is released. Your changes will go into OH3.1, then.

**Pull request merged**